Haskell/syntax_test_haskell.hs

```diff
@@ -93,21 +102,28 @@
 --
    a a = (+) a 2
 --     ^ keyword.operator.haskell
---       ^^^ entity.name.function.infix.haskell
+--       ^^^ variable.function.infix.haskell
+--        ^ keyword.operator.haskell
 --             ^ constant.numeric.integer.decimal.haskell
    a a = (-) a 2
 --     ^ keyword.operator.haskell
---       ^^^ entity.name.function.infix.haskell
+--       ^^^ variable.function.infix.haskell
 --             ^ constant.numeric.integer.decimal.haskell
    a a = (*) a 2
 --     ^ keyword.operator.haskell
---       ^^^ entity.name.function.infix.haskell
+--       ^^^ variable.function.infix.haskell
 --             ^ constant.numeric.integer.decimal.haskell
    a a = (/) a 2
 --     ^ keyword.operator.haskell
---       ^^^ entity.name.function.infix.haskell
+--       ^^^ variable.function.infix.haskell
 --             ^ constant.numeric.integer.decimal.haskell
 
+
+   a `member` x
+--   ^^^^^^^^ keyword.operator.function.infix.haskell
```

Comment:
This *definitely* shouldn't be `keyword` anything. This syntax is analogous to the prefix parentheses syntax for infix operators. IMO, the contents of the backticks should be `variable.function.prefix` and the backticks themselves should be just `punctuation` scope.

Reply:
Operators (`keyword.operator`) are behind the keyword top scope. We could probably make this `variable.function`, but I would need to think about this more considering how basically everything is a function and the backtick notation only turns a function into an infix function.

Reply:
I would generally consider `keyword` in any language to be from a statically fixed set. Anything which *isn't* from a statically fixed set probably shouldn't be a `keyword`. My two cents anyway.

Reply:
Seeing how most operators in functional languages (like Haskell) aren't proper keywords, are you saying an "operator" like `+` shouldn't be scoped `keyword.operator`? If so, what other scope should be used? `variable.function`?

How do other functional languages like Scheme, Lisp or Scala do it? Since you wrote the latter, I imagine you having spent more time with this problem than I did.

Reply:
In Scala, I just gave up and I don't use `variable.function` at all. Basically, there's no meaningful way to distinguish functions and operators in Scala, and in all but a very very small set of cases, they behave identically. Haskell does give a bit more information in some cases – specifically the backticks for infix function usage, and the parentheses for prefix oprator usage – the question is just whether or not we want to *use* that information and how we would use it.

Scoping either of them as `keyword` seems very much wrong. In my view, the best thing to do would be to scope the prefix paren syntax as `variable.function.infix`, which we could also use for operators in all positions if the lexical syntax is sufficiently disambiguated. Infix backtick syntax would be something like `variable.function.prefix`. Scoping all non-operator identifiers as `variable.function.prefix` would technically be *correct* in a sense, but it's also probably worthless and we shouldn't do that.

Reply:
Imo we *should* highlight infix function calls because they affect the syntax, but we can leave parenthesized infix operators to use them as prefix functions alone or do whatever we do for any other function. They should still be scoped as *something*, but not `variable.function`.

Now, to me, the question remains what infix functions should be scoped as. Semantically, they function like binary operators and this matches well with other languages. If the operator scope is a subscope of `keyword`, then so be it. Scoping them as `variable.function` is also an option, but I worry about color schemes not targeting that or assigning a style that is too subtle (my current scheme just makes them italics). I'm unsure how much bias I have regarding this.
Maybe I'll just try both and see how things work out.

Reply:
Well, at least to me, the backticks and paren syntaxes are really the same thing: each takes one fixity and swaps it to the other, and both affect syntax. Regarding scoping, I'm just quite worried that `keyword` is too much.